### PR TITLE
postgres: validate message syntax before parsing

### LINKED ISSUE #12340: [postgres_proxy] assert failure with untrusted buffer when onData()
I'm currently working on fuzz test(which generates random bytes for onData() and onWrite() to see whether we could crash the Envoy) for network-level filters.
When I was testing on postres_proxy filter with some untrusted data, an assert failure occurred inside linearize:
RELEASE_ASSERT(size <= length(), "Linearize size exceeds buffer size");
https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/network/postgres_proxy/postgres_decoder.cc#L201

This error only happens in fuzzer or when upstream server is on bad state, so it is not security-critical now.
But I think that we could deal with this error more gracefully. (So that we could make the filter more robust to upstream errors, and enable the fuzzer to continue testing it). 

My idea is that we could make it just like other invalid error handles in this file, which is `return false;`, before calling linearize?
This solution looks like this(from line 200):
  ```
auto bytesToRead = length - 4;
  if(bytesToRead>data.length()){
    return false;
  }
  message.assign(std::string(static_cast<char*>(data.linearize(bytesToRead)), bytesToRead));

```
This issue can be reproduced in unit test by adding a case as below(test/extensions/filters/network/postgres_proxy/postgres_decoder_test.cc):
```
TEST_P(PostgresProxyFrontendEncrDecoderTest, AssertFailure) {
  std::string str_data;
  for(int i=0;i<8;i++){
    str_data.push_back('\0');
  }
  Buffer::OwnedImpl data(str_data);
  decoder_->onData(data, false);
}
```
If anyone has a better idea, please share with me or make a pull request and link it here. Thanks! 
/cc @dio 
/cc @fabriziomello 
/cc @cpakulski

### PR DESCRIPTION
**Pull request by cpakulski**
Commit Message:
Validate postgres messages before parsing.

Additional Description:
Introduced InSync and OutOfSync states in decoder. When decoder detects a wrongly formatted message, it stops parsing and moves to OutOfSync state. Continuing parsing after encountering message with wrong syntax may lead to interpreting random bytes as length of the message and possibly causing OOM.

Risk Level: Low
Testing: Added unit tests and run full regression tests.
Docs Changes: No.
Release Notes: No.
Platform Specific Features:
Fixes #12340

### CODE REVIEW

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.h [hunk 1]
```diff
@@ -150,17 +160,20 @@ class DecoderImpl : public Decoder, Logger::Loggable<Logger::Id::filter> {
...
   DecoderCallbacks* callbacks_{};
   PostgresSession session_{};
 
   // The following fields store result of message parsing.
-  char command_{};
+  char command_{'-'};
   std::string message_;
   uint32_t message_len_{};
 
+  static const std::string FRONTEND;
```
Comment (lizan) [inline source/extensions/filters/network/postgres_proxy/postgres_decoder.h:174]: static const string (non-POD) is source of flaky test, use singleton (CONSTRUCT_ON_FIRST_USE) or constexpr string_view.

Comment (cpakulski) [reply source/extensions/filters/network/postgres_proxy/postgres_decoder.h]: Converted to `constexpr string_view`. Thanks!

**Comment by zuercher** [comment]
cc @fabriziomello could have a look as an extension owner?

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc [hunk 2]
```diff
@@ -176,134 +179,216 @@ void DecoderImpl::initialize() {
...
+  }
+
+  // Validate the message before processing.
+  const MsgBodyReader& f = std::get<1>(first_);
+  const auto msgParser = f();
+  // Run the validation.
+  message_len_ = data.peekBEInt<uint32_t>(0);
+// MAX_STARTUP_PACKET_LENGTH is defined in Postgres source code
+// as maximum size of initial packet.
+#define MAX_STARTUP_PACKET_LENGTH 10000
```
Comment (fabriziomello) [inline source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:223]: Perhaps move it to `postgres_decoder.h` and add the following link to comments: https://github.com/postgres/postgres/search?q=MAX_STARTUP_PACKET_LENGTH&type=code

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc [hunk 3]
```diff
@@ -176,134 +179,216 @@ void DecoderImpl::initialize() {
...
+   and indicates that session should be encrypted, the decoder stays in InitState, because the
+   initial message will be received again after transport socket negotiates SSL. If the message
+   syntax is incorrect, the decoder will move to OutOfSyncState, in which messages are not parsed.
+*/
+Decoder::Result DecoderImpl::onDataInit(Buffer::Instance& data, bool) {
+  ASSERT(state_ == State::InitState);
 
   // The minimum size of the message sufficient for parsing is 5 bytes.
-  if (data.length() < 5) {
+  if (data.length() < 4) {
```
Comment (fabriziomello) [inline source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:211]: What about create something like `#define MIN_MESSAGE_LENGTH 4` ? And seems the comment above is not in sync what code does...

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc [hunk 4]
```diff
@@ -176,134 +179,216 @@ void DecoderImpl::initialize() {
...
+  Decoder::Result result = Decoder::Result::ReadyForNext;
+  uint32_t code = data.peekBEInt<uint32_t>(4);
+  data.drain(4);
+  // Startup message with 1234 in the most significant 16 bits
+  // indicate request to encrypt.
+  if (code >= 0x04d20000) {
+    encrypted_ = true;
+    // Handler for SSLRequest (Int32(80877103) = 0x04d2162f)
+    // See details in https://www.postgresql.org/docs/current/protocol-message-formats.html.
+    if (code == 0x04d2162f) {
```
Comment (fabriziomello) [inline source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:253]: Both lines 249 and 253 use SSLRequest codes... for code legibility what about add some macros?

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc [hunk 5]
```diff
@@ -176,134 +179,216 @@ void DecoderImpl::initialize() {
...
-      callbacks_->incMessagesFrontend();
-      if (encrypted_) {
-        ENVOY_LOG(trace, "postgres_proxy: detected encrypted traffic.");
-        incSessionsEncrypted();
-        startup_ = false;
-      }
-      data.drain(data.length());
-      return encrypted_ ? Decoder::ReadyForNext : Decoder::Stopped;
+  Decoder::Result result = Decoder::Result::ReadyForNext;
+  uint32_t code = data.peekBEInt<uint32_t>(4);
```
Comment (fabriziomello) [inline source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:245]: Rename `code` to `startup_packet` just to use the Postgres vocabulary.

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc [hunk 6]
```diff
@@ -176,134 +179,216 @@ void DecoderImpl::initialize() {
...
+      encrypted_ = callbacks_->onSSLRequest();
+    }
+
+    // Count it as recognized frontend message.
+    callbacks_->incMessagesFrontend();
+    if (encrypted_) {
+      ENVOY_LOG(trace, "postgres_proxy: detected encrypted traffic.");
+      incSessionsEncrypted();
+      state_ = State::EncryptedState;
+      startup_ = false;
```
Comment (fabriziomello) [inline source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:268]: Now with the new `state_` control seems this flag is useless, no?

**Comment by fabriziomello** [comment]
> cc @fabriziomello could have a look as an extension owner?

Thanks for remember me ... did some minor comments. Me and @cpakulski already discussed a bit about this in PVT before he send the PR. Mostly about `MAX_STARTUP_PACKET_LENGTH` because we decided implement it the same way Postgres does.

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc (hunk 6 above)
Comment (cpakulski) [reply source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:268]: Yes. That is not used anymore. Thanks.

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc (hunk 5 above)
Comment (cpakulski) [reply source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:245]: Yeah, but this is not really a packet itself, but just 4-bytes code which can be either a version or encryption type.

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc (hunk 4 above)
Comment (cpakulski) [reply source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:253]: IMHO macro will make it more difficult to read and will not add any value. The meaning of this value is documented pretty clearly here.

**Comment by fabriziomello** [comment]
@cpakulski the [Postgres Regression](https://github.com/fabriziomello/envoy-postgres-regression/actions/runs/937433479) using your last commit [60b5585](https://github.com/envoyproxy/envoy/pull/16575/commits/60b5585ce368aa89f67bcf9df505a52258bde37a) worked but seems our current pipeline doesn't like it so much.

**Comment by cpakulski** [comment]
@fabriziomello Would you mind running Postgres Regression test on the latest commit? Thanks.

**Comment by fabriziomello** [comment]
> @fabriziomello Would you mind running Postgres Regression test on the latest commit? Thanks.

I did it when I saw your new commit and it ran without any issue: https://github.com/fabriziomello/envoy-postgres-regression/actions/runs/940000625

**Comment by cpakulski** [comment]
@lizan How can we move this PR forward? All regression tests have passed. 
If there are no style comments, it is ready for merge. Thanks!

**Review by lizan — APPROVED** [review]

**Pull request merged**